src/main/java/com/google/api/gax/grpc/PageContext.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.Iterables;
+import java.util.Iterator;
+
+public class PageContext<RequestT, ResponseT, ResourceT> {
+
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor;
+  private final RequestT request;
+  private final CallContext context;
+  private final ResponseT response;
+
+  public PageContext(
+      UnaryCallable<RequestT, ResponseT> callable,
+      PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+      RequestT request,
+      CallContext context) {
+    this.callable = callable;
+    this.pageDescriptor = pageDescriptor;
+    this.request = request;
+    this.context = context;
+
+    // Make the API call eagerly
+    this.response =
+        ApiExceptions.callAndTranslateApiException(callable.futureCall(request, context));
```

Comment:
If you have heavy initialization, it's a better practice to do the following:

* Make the constructor private
* Make a create method which calls the constructor
* Move the heavy initialization outside of the constructor, into the static create method
* Make any final setter calls on the constructed object with the heavy initialization result before returning

I do see that this came from PageImpl, but I didn't notice it before...